src/librustc/infer/lattice.rs

```diff
@@ -64,14 +65,14 @@ pub fn super_lattice_tys<'a, 'gcx, 'tcx, L>(this: &mut L,
     match (&a.sty, &b.sty) {
         (&ty::TyInfer(TyVar(..)), &ty::TyInfer(TyVar(..)))
             if infcx.type_var_diverges(a) && infcx.type_var_diverges(b) => {
-            let v = infcx.next_diverging_ty_var();
+            let v = infcx.next_diverging_ty_var(TypeVariableOrigin::LatticeVariable);
```

Comment:
We can give more information about lattice variables -- at minimum, we can give a span. The `this` variable is always of a type (either `Lub` or `Glb`) that has access to a `CombineFields` (`self.fields`, in each case). `CombineFields` has a field `trace` of type `TypeTrace`, and hiding in there is a field `cause` of type `ObligationCause`. This is basically "why we are computing the common-supertype or common-subtype of these types". At minimum though it'd be nice to add a method `fn cause(&self) -> &OlbligationCause<'tcx>` to `LatticeDir` that returns `&self.fields.cause`. Then we can create give a span here like `LatticeVariable(this.cause().span)`.